library/std/src/os/windows/dbg.rs

```diff
@@ -0,0 +1,130 @@
+/// if the debugger is attached, and is skipped over otherwise.
+///
+/// Returns `1` if the breakpoint was handled by the debugger,
+/// and `0` is returned if the breakpoint was handled by SEH.
+pub fn dbg_breakpoint() -> i32 {
+    // SAFETY: the call does not access any state shared between threads.
+    unsafe { __dbg_breakpoint() }
+}
+
+#[cfg(all(target_os = "windows", target_arch = "x86_64"))]
+core::arch::global_asm!(
+    r#"
+        .pushsection    .text
+
+        .globl          __dbg_breakpoint_flt
+        .p2align        4
+    __dbg_breakpoint_flt:
+        mov             eax, 1     # EXCEPTION_EXECUTE_HANDLER
+        ret
+
+        .globl          __dbg_breakpoint
+        .p2align        4
+        .def            __dbg_breakpoint; .scl 2; .type 32; .endef
+        .seh_proc       __dbg_breakpoint
+    __dbg_breakpoint:
+        sub             rsp, 64
+        .seh_stackalloc 64
+        .seh_endprologue
+    1:
+        int3
+        mov             eax, 1
+    2:
+    3:
+        add             rsp, 64
+        ret
+    4:
+        mov             eax, 0
+        jmp             3b
+        .seh_handler    __C_specific_handler, @except
+        .seh_handlerdata
+        .long 1                             # One handler entry
+        .long (1b)@IMGREL                   # Start address of __try block
+        .long (2b)@IMGREL                   # End address of __try block
+        .long (__dbg_breakpoint_flt)@IMGREL # Exception filter
+        .long (4b)@IMGREL                   # Exception handler
+        .text
+        .seh_endproc
+        .popsection
+    "#
+);
+
+#[cfg(all(target_os = "windows", target_arch = "aarch64"))]
+core::arch::global_asm!(
+    r#"
+        .pushsection    .text
+
+        .globl          __dbg_breakpoint_flt
+        .p2align        2
+    __dbg_breakpoint_flt:
+        mov             w0, 1     // EXCEPTION_EXECUTE_HANDLER
+        ret
+
+        .globl          __dbg_breakpoint
+        .p2align        2
+        .def            __dbg_breakpoint; .scl 2; .type 32; .endef
+        .seh_proc       __dbg_breakpoint
+    __dbg_breakpoint:
+        str             lr, [sp, #-16]!
+        .seh_save_reg_x lr, 16
+        .seh_endprologue
+    1:
+        brk             #0xf000
+        mov             w0, 1
+    2:
+    3:
+        .seh_startepilogue
+        ldr             lr, [sp], #16
+        .seh_save_reg_x lr, 16
+        .seh_endepilogue
+        ret
+    4:
+        mov             w0, wzr
+        b               3b
+        .seh_handler    __C_specific_handler, @except
+        .seh_handlerdata
+        .long 1                             // One handler entry
+        .long (1b)@IMGREL                   // Start address of __try block
+        .long (2b)@IMGREL                   // End address of __try block
+        .long (__dbg_breakpoint_flt)@IMGREL // Exception filter
+        .long (4b)@IMGREL                   // Exception handler
```

Comment:
Where did you copy this code from?

Reply:
The SEH directives are from the output of `clang -c -fms-extensions -O3 -S seh.c` where `seh.c` is
```c
__declspec(noinline) int try_break_into_debugger()
{
 __try
  {
    __debugbreak();
    return 0;
  }
  __except (1)
  {
    return 1;
  }
}

int main() {
  try_break_into_debugger();
  return 0;
}
```

The assembly was a tweaked version of the output, the rest creates the SEH entries.